File: AlgoExpert/2. Medium/C++/Three Number Sum/Three Number Sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// O(n^2) time | O(n) space

vector<vector<int>> threeNumberSum(vector<int> array, int target) {
	sort(array.begin(), array.end());
	vector<vector<int>> triplets;
	for (int i = 0; i < array.size() - 2; i++) {
		int left = i + 1;
		int right = array.size() - 1;
		while (left < right) {
			int currentSum = array[i] + array[left] + array[right];
			if (currentSum == targetSum) {
				triplets.push_back({array[i], array[left], array[right]});
				left++;
				right--;
			} else if (currentSum < targetSum) {
				left++;
			} else if (currentSum > targetSum) {
				right--;
			}
		}
	}
	return triplets;
}

File: AlgoExpert/2. Medium/Python/Three Number Sum/Three Number Sum.py
# O(n^2) time | O(n) space

def threeNumberSum(array, targetSum):
	array.sort()
	triplets = []
	for i in range(len(array) - 2):
		left = i + 1
		right = len(array) - 1
		while left < right:
			currentSum = array[i] + array[left] + array[right]
			if currentSum == targetSum:
				triplets.append([array[i], array[left], array[right])
				left += 1
				right -= 1
			elif currentSum < targetSum:
				left += 1
			elif currentSum > targetSum:
				right -= 1
	return triplets